pytests/test_operators.py

```diff
@@ -262,3 +338,137 @@ def running_count(type_to_count, event):
         ),
     ):
         run_main(flow)
+
+
+def test_reduce_window(recovery_config):
+    flow = Dataflow()
+
+    inp = [
+        ("ALL", 1),
+        ("ALL", 1),
+        "BOOM",
+        ("ALL", 1),
+        ("ALL", 1),
+    ]
+    flow.input("inp", TestingInputConfig(inp))
+
+    armed = Event()
+    armed.set()
+
+    def trigger(item):
+        if item == "BOOM":
+            if armed.is_set():
+                raise RuntimeError("BOOM")
+            else:
+                return []
+        else:
+            return [item]
+
+    flow.flat_map(trigger)
+
+    start_at = datetime(2022, 1, 1)
+    # This will result in times for events of +0, +4, +8, +12.
+    clock_config = TestingClockConfig(item_incr=timedelta(seconds=4), start_at=start_at)
+    # And since the window is +10, we should get a window with value
+    # of 3 and then 1.
+    window_config = TumblingWindowConfig(
+        length=timedelta(seconds=10), start_at=start_at
+    )
+
+    def add(acc, x):
+        return acc + x
+
+    flow.reduce_window("add", clock_config, window_config, add)
+
+    out = []
+    flow.capture(TestingOutputConfig(out))
+
+    with raises(RuntimeError):
+        run_main(flow, epoch_config=epoch_config, recovery_config=recovery_config)
+
+    # No windows yet after epoch 2.
+    assert sorted(out) == sorted([])
+
+    # Disable bomb
+    armed.clear()
+    out.clear()
+
+    # Recover
+    run_main(flow, epoch_config=epoch_config, recovery_config=recovery_config)
+
+    # But it remembers the first two items in the first window.
+    assert sorted(out) == sorted([("ALL", 3), ("ALL", 1)])
+
+
+def test_fold_window(recovery_config):
+    flow = Dataflow()
+
+    # Remember that clocks are built per-key so the `TestingClock` in
```

Comment:
Probably a smell but let's table that for now 🤔

Reply:
This is sort of changed in https://github.com/bytewax/bytewax/pull/139.

Not the per-key thing, but it'll be less confusing? Although I'm having some trouble with the `TestingClockConfig` in another PR, so the smell does remain...

Reply:
I figured out what was confusing me. It's that we were recovering and using a half-consumed generator. https://github.com/bytewax/bytewax/pull/144 has the tests out of this working.